### PR TITLE
Sync tests by summary name

### LINKED ISSUE #5: automatic TestSet and xTest creation
Please add optional functionality for automatic creation of "Test Set" and "xTest" issues in Jira.
For every test class annotated with @XrayTestSet a "Test Set" should be created if missing.
For every test method annotated with @Test a "xTest" should be created if missing, then added to corresponding “Test Set”. If xTest was found, should be only added to “Test Set”.
After test run, every xTest in "Test Set" that wasn't updated (wasn't seen in @XrayTestSet) should be removed from "Test Set". If test isn’t used anymore, it should be tagged as obsolete - such tests could be removed manually some day when Jira allows that functionality.

Automatic removal of "Test Set" should not be implemented.

### PR DESCRIPTION
**Pull request by radzimir**
---
name: Pull Request
about: Create a pull request for your changes
title: 'automatic creation of TestSet, Test-s, TestRun'
labels: enhancement
assignees: ''

---

# Description

Automatic creation of TestSet, Test-s, TestRun. All work is based on the original branch from Mike.

Fixes #5 

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/jql/predefined/IssueType.java [hunk 1]
```diff
@@ -27,23 +27,29 @@
...
     @Override
     public String toString() {
-        return issueType.getName();
+        return getIssueType().getName();
     }
 
+    // Properties can be set later after class loading, so resolve issue type at runtime
     public JiraIssueType getIssueType() {
-        return this.issueType;
+        switch ( this ) {
```
Comment (Zsar) [inline src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/jql/predefined/IssueType.java]: FWIW: You can get around the `default` and its assorted maintenance chore by making each enum value its own anonymous subclass of `IssueType`. Using an interface declaring `getIssueType` implemented by `IssueType` each new value will cause a compile time error if it does not implement `getIssueType` => no chance to forget an implementation and actually do throw.

... Admittedly that does look a little ugly, but at least not at runtime.

Comment (radzimir) [reply src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/jql/predefined/IssueType.java]: OK, Commited as e1439d2d66a7d477cd2d

File: src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java [hunk 2]
```diff
@@ -414,17 +435,17 @@ public XrayTestExecutionImport(XrayTestExecutionIssue testExecutionIssue) {
         this.info.setSummary(testExecutionIssue.getSummary());
         this.info.version = testExecutionIssue.getFixVersions().stream().findFirst().map(JiraNameReference::getName).orElse(null);
         this.info.revision = testExecutionIssue.getRevision();
-        this.info.user = testExecutionIssue.getAssignee().getName();
+        this.info.user = testExecutionIssue.getAssignee() != null ? testExecutionIssue.getAssignee().getName() : null;
```
Comment (Zsar) [inline src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java]: `JiraIssue#getAssignee` calls `JiraIssue#getOrCreateEntity` and cannot return `null`.

File: src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java [hunk 3]
```diff
@@ -414,17 +435,17 @@ public XrayTestExecutionImport(XrayTestExecutionIssue testExecutionIssue) {
...
         this.info.startDate = testExecutionIssue.getStartDate();
         this.info.finishDate = testExecutionIssue.getFinishDate();
 
         List<String> testPlanKeys = testExecutionIssue.getTestPlanKeys();
-        if  (testPlanKeys.size() > 0) {
+        if  (testPlanKeys != null && testPlanKeys.size() > 0) {
```
Comment (Zsar) [inline src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java]: `XrayTestExecutionIssue#getTestPlanKeys` calls `JiraIssue#getOrCreateFieldList` and cannot return `null`.

File: src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java [hunk 4]
```diff
@@ -414,17 +435,17 @@ public XrayTestExecutionImport(XrayTestExecutionIssue testExecutionIssue) {
...
             this.info.testPlanKey = testPlanKeys.get(0);
         }
 
         List<String> testEnvironments = testExecutionIssue.getTestEnvironments();
-        if (testEnvironments.size() > 0) {
+        if (testEnvironments != null && testEnvironments.size() > 0) {
```
Comment (Zsar) [inline src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java]: `XrayTestExecutionIssue#getTestEnvironments` calls `JiraIssue#getOrCreateFieldList` and cannot return `null`.

File: src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/jql/predefined/IssueType.java [hunk 5]
```diff
@@ -26,24 +26,36 @@
...
 
-public enum IssueType {
-    Test(new JiraIssueType(PropertyManager.getProperty(XrayMapper.PROPERTY_TEST_ISSUETYPE_NAME, "Test"))),
-    TestExecution(new JiraIssueType(PropertyManager.getProperty(XrayMapper.PROPERTY_TEST_EXECUTION_ISSUETYPE_NAME, "Test Execution"))),
-    TestSet(new JiraIssueType(PropertyManager.getProperty(XrayMapper.PROPERTY_TEST_SET_ISSUETYPE_NAME,"Test Set"))),
-    ;
+import java.util.function.Supplier;
 
-    private final JiraIssueType issueType;
+public enum IssueType implements Supplier<JiraIssueType> {
```
Comment (Zsar) [inline src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/jql/predefined/IssueType.java:31]: Ah! Nice one.

File: src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/synchronize/AbstractXrayResultsSynchronizer.java [hunk 6]
```diff
@@ -128,18 +117,23 @@ private XrayTestExecutionIssue getTestExecutionIssue() {
             final ExecutionContext executionContext = ExecutionContextController.getCurrentExecutionContext();
             xrayMapper.updateTestExecution(testExecutionIssue, executionContext);
 
-            final Optional<XrayTestExecutionIssue> optionalExistingTestExecution = Optional.ofNullable(xrayMapper.queryTestExecution(testExecutionIssue))
-                    .flatMap(jqlQuery -> xrayUtils.searchIssues(jqlQuery, XrayTestExecutionIssue::new).findFirst());
+            JqlQuery queryTestExecution = xrayMapper.queryTestExecution( testExecutionIssue );
+            List<XrayTestExecutionIssue> testExecutionIssueList = xrayUtils.searchIssues( queryTestExecution, XrayTestExecutionIssue::new )
+                    .collect( Collectors.toList() );
+            if( testExecutionIssueList.size() > 1) {
```
Comment (Zsar) [inline src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/synchronize/AbstractXrayResultsSynchronizer.java:123]: `xrayUtils.searchIssues(queryTestExecution, XrayTestExecutionIssue::new).collect(MoreCollectors.toOptional())` ?
(using `com.google.common.collect.MoreCollectors`)

**Review by Zsar — CHANGES_REQUESTED** [review]
Am too fresh to judge sensibility of the changes, but here are a few "basic Java" things I noticed. Going to defer to @martingrossmann with regards to business logic.

File: src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java (hunk 2 above)
Comment (radzimir) [reply src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java]: agree, fixing now

File: src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java (hunk 3 above)
Comment (radzimir) [reply src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java]: agree, fixing now

File: src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java (hunk 4 above)
Comment (radzimir) [reply src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/mapper/xray/XrayTestExecutionImport.java]: agree, fixing now

File: src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/synchronize/AbstractXrayResultsSynchronizer.java (hunk 6 above)
Comment (radzimir) [reply src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/synchronize/AbstractXrayResultsSynchronizer.java:123]: It is possible that JQL Query matches mor than one Test Issue, in that case it is an error that need to be communicated fixed by extending the query.
MoreCollectors.toOptional() do indeed throws an Exception if stream is empty or contains more elements, but I find it more readable if we do that check in our code.
I suggest to keep it as it is.

Comment (Zsar) [reply src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/synchronize/AbstractXrayResultsSynchronizer.java:123]: Fair.

**Comment by martingrossmann** [comment]
Close this PR because it was replaced by #11

**Pull request closed without merging**